nima/webclient/webclient/src/main/java/io/helidon/nima/webclient/Proxy.java

```diff
@@ -0,0 +1,528 @@
+    private static final Pattern IP_V6_IDENTIFIER = Pattern.compile("^\\[(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}]$");
+    private static final Pattern IP_V6_HEX_IDENTIFIER = Pattern
+            .compile("^\\[((?:[0-9A-Fa-f]{1,4}(?::[0-9A-Fa-f]{1,4})*)?)::((?:[0-9A-Fa-f]{1,4}(?::[0-9A-Fa-f]{1,4})*)?)]$");
+    private static final Pattern IP_V6_HOST = Pattern.compile("^(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$");
+    private static final Pattern IP_V6_HEX_HOST = Pattern
+            .compile("^((?:[0-9A-Fa-f]{1,4}(?::[0-9A-Fa-f]{1,4})*)?)::((?:[0-9A-Fa-f]{1,4}(?::[0-9A-Fa-f]{1,4})*)?)$");
+
+    private static final LruCache<String, Boolean> IVP6_HOST_MATCH_RESULTS = LruCache.<String, Boolean>builder()
+            .capacity(100)
+            .build();
+    private static final LruCache<String, Boolean> IVP6_IDENTIFIER_MATCH_RESULTS = LruCache.<String, Boolean>builder()
+            .capacity(100)
+            .build();
+
+    private final ProxyType type;
+    private final String host;
+    private final int port;
+    private final Function<URI, Boolean> noProxy;
+    private final Optional<String> username;
+    private final Optional<char[]> password;
+
+    private Proxy(Proxy.Builder builder) {
+        this.type = builder.type();
+        this.host = builder.host();
+
+        this.port = builder.port();
+        this.username = builder.username();
+        this.password = builder.password();
+        this.noProxy = prepareNoProxy(builder.noProxyHosts());
+    }
+
+    /**
+     * Fluent API builder for new instances.
+     *
+     * @return a new builder
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * A Proxy instance that does not proxy requests.
+     *
+     * @return a new instance with no proxy definition
+     */
+    public static Proxy noProxy() {
+        return NO_PROXY;
+    }
+
+    /**
+     * Create a new proxy instance from configuration.
+     * {@code
+     * proxy:
+     * http:
+     * uri: https://www.example.org
+     * https:
+     * uri: https://www.example.org
+     * no-proxy: ["*.example.org", "localhost"]
+     * }
+     *
+     * @param config configuration, should be located on a key that has proxy as a subkey
+     * @return proxy instance
+     */
+    public static Proxy create(Config config) {
+        return builder()
+                .config(config)
+                .build();
+    }
+
+    /**
+     * Verifies whether the current host is inside noHosts.
+     *
+     * @param uri the uri
+     * @return true if it is in no hosts, otherwise false
+     */
+    public boolean isNoHosts(URI uri) {
+        return noProxy.apply(uri);
+    }
+
+    /**
+     * Get proxy type. For testing purposes.
+     *
+     * @return the proxy type
+     */
+    ProxyType type() {
+        return type;
+    }
+
+    static Function<URI, Boolean> prepareNoProxy(Set<String> noProxyHosts) {
```

Comment:
Please use UriHelper instead of URI in this function.

Reply:
Done